packages/ui/src/components/masonry.tsx

```diff
@@ -0,0 +1,148 @@
+"use client";
+
+import React, { useEffect, useState } from "react";
+
+// NOTE: Source code from https://github.com/askides/react-plock/blob/1581be9051f1a355104570864d6ec02674cf6809/libs/react-plock/src/Plock.tsx
+export function useMediaValues(
+  medias: number[] | undefined,
+  columns: number[],
+  gap: number[]
+) {
+  const [values, setValues] = useState({ columns: 0, gap: 1 });
+
+  useEffect(() => {
+    if (!medias) {
+      setValues({ columns: columns[0], gap: gap[0] });
+      return;
+    }
+
+    const mediaQueries = medias.map((media) =>
+      window.matchMedia(`(min-width: ${media}px)`)
+    );
+
+    const onSizeChange = () => {
+      let matches = 0;
+
+      mediaQueries.forEach((mediaQuery) => {
+        if (mediaQuery.matches) {
+          matches++;
+        }
+      });
+
+      // Update Values
+      const idx = Math.min(mediaQueries.length - 1, Math.max(0, matches));
+      setValues({ columns: columns[idx], gap: gap[idx] });
+    };
+
+    // Initial Call
+    onSizeChange();
+
+    // Apply Listeners
+    for (const mediaQuery of mediaQueries) {
+      mediaQuery.addEventListener("change", onSizeChange);
+    }
+
+    return () => {
+      for (const mediaQuery of mediaQueries) {
+        mediaQuery.removeEventListener("change", onSizeChange);
+      }
+    };
+  }, [values.columns, values.gap]);
+
+  return values;
+}
+
+export type MasonryProps<T> = React.ComponentPropsWithoutRef<"div"> & {
+  items: T[];
+  render: (item: T, idx: number) => React.ReactNode;
+  config: {
+    columns: number | number[];
+    gap: number | number[];
+    media?: number[];
+  };
+};
+
+export function createSafeArray(data: number | number[]) {
+  return Array.isArray(data) ? data : [data];
+}
+
+export function Masonry<T>({
```

Comment:
Added a masonry list ui component in the UI package. The component is from the package called react-plock https://github.com/askides/react-plock.